File: packages/shared/src/transport/frame/Codec.ts
import type { ErrorCode } from "../../protocol/Error.js";
import { createError } from "../../protocol/Error.js";

const PACKET_HEADER_SIZE = 4;
const FRAME_HEADER_SIZE = 6;
const OPEN_FIXED_SIZE = 6;
const WINDOW_PAYLOAD_SIZE = 4;
const RESET_CODE_SIZE = 2;
const UINT16_MAX = 0xffff;
const UINT32_MAX = 0xffff_ffff;

const decoder = new TextDecoder("utf-8", { fatal: true });
const encoder = new TextEncoder();

export const FRAME_PROTOCOL_VERSION = 1;
export const FRAME_MAX_DATA_SIZE = 64 * 1024;
export const FRAME_MAX_OPEN_METADATA_SIZE = 64 * 1024;
export const FRAME_MAX_OPEN_SERVICE_SIZE = 256;
export const TRANSPORT_MAX_FRAME_SIZE = 16 * 1024 * 1024;

export const frameResetCodes = {
    unsupportedService: 1,
    serviceRejected: 2,
    serviceFailed: 3,
    cancelled: 4,
    streamProtocolError: 5,
} as const;

export type FrameResetCode =
    (typeof frameResetCodes)[keyof typeof frameResetCodes];

export type Frame =
    | {
          readonly type: "open";
          readonly streamId: number;
          readonly receiveWindow: number;
          readonly service: string;
          readonly metadata: Uint8Array;
      }
    | {
          readonly type: "data";
          readonly streamId: number;
          readonly data: Uint8Array;
      }
    | {
          readonly type: "window";
          readonly streamId: number;
          readonly creditDelta: number;
      }
    | { readonly type: "fin"; readonly streamId: number }
    | {
          readonly type: "reset";
          readonly streamId: number;
          readonly code: number;
          readonly message: string;
      };

enum FrameType {
    Open = 0x01,
    Data = 0x02,
    Window = 0x03,
    Fin = 0x04,
    Reset = 0x05,
}

/**
 * @compat packet-buffer-pre-frame-v1
 * @removeAt 0.7.10
 * Generic length-prefixed packet decoder retained for protocols that have not
 * migrated to Frame v1 yet. The packet body has no Frame semantics.
 */
export class PacketBuffer {
    readonly #maxPacketSize: number;
    #buffer: Uint8Array<ArrayBufferLike> = new Uint8Array();
    #end = 0;
    #start = 0;

    constructor(maxPacketSize = TRANSPORT_MAX_FRAME_SIZE) {
        this.#maxPacketSize = maxPacketSize;
    }

    get empty(): boolean {
        return this.#start === this.#end;
    }

    push(chunk: Uint8Array): Uint8Array[] {
        if (chunk.byteLength === 0) return [];
        this.#append(chunk);
        const packets: Uint8Array[] = [];
        while (this.#end - this.#start >= PACKET_HEADER_SIZE) {
            const payloadLength = readU32(this.#buffer, this.#start);
            assertPacketSize(payloadLength, this.#maxPacketSize);
            const packetLength = PACKET_HEADER_SIZE + payloadLength;
            if (this.#end - this.#start < packetLength) break;
            const payloadStart = this.#start + PACKET_HEADER_SIZE;
            packets.push(
                this.#buffer.slice(payloadStart, this.#start + packetLength),
            );
            this.#start += packetLength;
        }
        if (this.#start === this.#end) {
            this.#buffer = new Uint8Array();
            this.#start = 0;
            this.#end = 0;
        }
        return packets;
    }

    reset(): void {
        this.#buffer = new Uint8Array();
        this.#start = 0;
        this.#end = 0;
    }

    #append(chunk: Uint8Array): void {
        const unread = this.#end - this.#start;
        const required = unread + chunk.byteLength;
        if (this.#buffer.byteLength < required) {
            let capacity = Math.max(1024, this.#buffer.byteLength);
            while (capacity < required) capacity *= 2;
            const next = new Uint8Array(capacity);
            if (unread > 0)
                next.set(this.#buffer.subarray(this.#start, this.#end), 0);
            this.#buffer = next;
            this.#start = 0;
            this.#end = unread;
        } else if (this.#end + chunk.byteLength > this.#buffer.byteLength) {
            this.#buffer.copyWithin(0, this.#start, this.#end);
            this.#start = 0;
            this.#end = unread;
        }
        this.#buffer.set(chunk, this.#end);
        this.#end += chunk.byteLength;
    }
}

export function encodePacket(
    payload: Uint8Array,
    maxPacketSize = TRANSPORT_MAX_FRAME_SIZE,
): Uint8Array {
    assertPacketSize(payload.byteLength, maxPacketSize);
    const packet = new Uint8Array(PACKET_HEADER_SIZE + payload.byteLength);
    writeU32(packet, 0, payload.byteLength);
    packet.set(payload, PACKET_HEADER_SIZE);
    return packet;
}

export function decodePacket(
    packet: Uint8Array,
    maxPacketSize = TRANSPORT_MAX_FRAME_SIZE,
): Uint8Array {
    if (packet.byteLength < PACKET_HEADER_SIZE) {
        throw protocolError("Packet header is incomplete.");
    }
    const payloadLength = readU32(packet, 0);
    assertPacketSize(payloadLength, maxPacketSize);
    if (packet.byteLength !== PACKET_HEADER_SIZE + payloadLength) {
        throw protocolError("Packet length does not match payload length.");
    }
    return packet.slice(PACKET_HEADER_SIZE);
}

export class FrameBuffer {
    readonly #packets: PacketBuffer;

    constructor(maxFrameSize = TRANSPORT_MAX_FRAME_SIZE) {
        this.#packets = new PacketBuffer(maxFrameSize);
    }

    get empty(): boolean {
        return this.#packets.empty;
    }

    push(chunk: Uint8Array): Frame[] {
        return this.#packets.push(chunk).map(decodeFramePayload);
    }

    reset(): void {
        this.#packets.reset();
    }
}

export function encodeFrame(frame: Frame): Uint8Array {
    assertStreamId(frame.streamId);
    let payload: Uint8Array;
    switch (frame.type) {
        case "open":
            payload = encodeOpen(frame);
            break;
        case "data":
            if (frame.data.byteLength > FRAME_MAX_DATA_SIZE) {
                throw protocolError(
                    `DATA payload exceeds ${FRAME_MAX_DATA_SIZE} bytes.`,
                    "protocol.frameTooLarge",
                );
            }
            payload = frame.data;
            break;
        case "window":
            assertPositiveU32(frame.creditDelta, "WINDOW credit");
            payload = new Uint8Array(WINDOW_PAYLOAD_SIZE);
            writeU32(payload, 0, frame.creditDelta);
            break;
        case "fin":
            payload = new Uint8Array();
            break;
        case "reset": {
            assertU16(frame.code, "RESET code");
            const message = encoder.encode(frame.message);
            payload = new Uint8Array(RESET_CODE_SIZE + message.byteLength);
            writeU16(payload, 0, frame.code);
            payload.set(message, RESET_CODE_SIZE);
            break;
        }
    }

    const body = new Uint8Array(FRAME_HEADER_SIZE + payload.byteLength);
    body[0] = FRAME_PROTOCOL_VERSION;
    body[1] = typeCode(frame.type);
    writeU32(body, 2, frame.streamId);
    body.set(payload, FRAME_HEADER_SIZE);
    return encodePacket(body);
}

export function decodeFrame(packet: Uint8Array): Frame {
    return decodeFramePayload(decodePacket(packet));
}

function encodeOpen(frame: Extract<Frame, { type: "open" }>): Uint8Array {
    assertPositiveU32(frame.receiveWindow, "OPEN receive window");
    const service = encoder.encode(frame.service);
    if (
        service.byteLength === 0 ||
        service.byteLength > UINT16_MAX ||
        service.byteLength > FRAME_MAX_OPEN_SERVICE_SIZE
    ) {
        throw protocolError("OPEN service must be a non-empty UTF-8 name.");
    }
    if (frame.metadata.byteLength > FRAME_MAX_OPEN_METADATA_SIZE) {
        throw protocolError(
            `OPEN metadata exceeds ${FRAME_MAX_OPEN_METADATA_SIZE} bytes.`,
            "protocol.frameTooLarge",
        );
    }
    const payload = new Uint8Array(
        OPEN_FIXED_SIZE + service.byteLength + frame.metadata.byteLength,
    );
    writeU32(payload, 0, frame.receiveWindow);
    writeU16(payload, 4, service.byteLength);
    payload.set(service, OPEN_FIXED_SIZE);
    payload.set(frame.metadata, OPEN_FIXED_SIZE + service.byteLength);
    return payload;
}

function decodeFramePayload(body: Uint8Array): Frame {
    if (body.byteLength < FRAME_HEADER_SIZE) {
        throw protocolError("Frame header is incomplete.");
    }
    const version = body[0];
    if (version !== FRAME_PROTOCOL_VERSION) {
        throw protocolError(`Unsupported Frame protocol version ${version}.`);
    }
    const type = body[1];
    const streamId = readU32(body, 2);
    assertStreamId(streamId);
    const payload = body.subarray(FRAME_HEADER_SIZE);

    switch (type) {
        case FrameType.Open:
            return decodeOpen(streamId, payload);
        case FrameType.Data:
            if (payload.byteLength > FRAME_MAX_DATA_SIZE) {
                throw protocolError(
                    `DATA payload exceeds ${FRAME_MAX_DATA_SIZE} bytes.`,
                    "protocol.frameTooLarge",
                );
            }
            return { type: "data", streamId, data: Uint8Array.from(payload) };
        case FrameType.Window: {
            if (payload.byteLength !== WINDOW_PAYLOAD_SIZE) {
                throw protocolError("WINDOW payload must be exactly 4 bytes.");
            }
            const creditDelta = readU32(payload, 0);
            assertPositiveU32(creditDelta, "WINDOW credit");
            return { type: "window", streamId, creditDelta };
        }
        case FrameType.Fin:
            if (payload.byteLength !== 0) {
                throw protocolError("FIN payload must be empty.");
            }
            return { type: "fin", streamId };
        case FrameType.Reset: {
            if (payload.byteLength < RESET_CODE_SIZE) {
                throw protocolError("RESET payload is incomplete.");
            }
            const code = readU16(payload, 0);
            assertU16(code, "RESET code");
            return {
                type: "reset",
                streamId,
                code,
                message: decodeUtf8(
                    payload.subarray(RESET_CODE_SIZE),
                    "RESET message",
                ),
            };
        }
        default:
            throw protocolError(`Unknown Frame type ${String(type)}.`);
    }
}

function decodeOpen(streamId: number, payload: Uint8Array): Frame {
    if (payload.byteLength < OPEN_FIXED_SIZE) {
        throw protocolError("OPEN payload is incomplete.");
    }
    const receiveWindow = readU32(payload, 0);
    assertPositiveU32(receiveWindow, "OPEN receive window");
    const serviceLength = readU16(payload, 4);
    if (serviceLength === 0) {
        throw protocolError("OPEN service must not be empty.");
    }
    if (serviceLength > FRAME_MAX_OPEN_SERVICE_SIZE) {
        throw protocolError(
            `OPEN service exceeds ${FRAME_MAX_OPEN_SERVICE_SIZE} bytes.`,
            "protocol.frameTooLarge",
        );
    }
    const serviceEnd = OPEN_FIXED_SIZE + serviceLength;
    if (serviceEnd > payload.byteLength) {
        throw protocolError("OPEN service length exceeds payload length.");
    }
    const service = decodeUtf8(
        payload.subarray(OPEN_FIXED_SIZE, serviceEnd),
        "OPEN service",
    );
    const metadataLength = payload.byteLength - serviceEnd;
    if (metadataLength > FRAME_MAX_OPEN_METADATA_SIZE) {
        throw protocolError(
            `OPEN metadata exceeds ${FRAME_MAX_OPEN_METADATA_SIZE} bytes.`,
            "protocol.frameTooLarge",
        );
    }
    return {
        type: "open",
        streamId,
        receiveWindow,
        service,
        metadata: payload.slice(serviceEnd),
    };
}

function typeCode(type: Frame["type"]): number {
    switch (type) {
        case "open":
            return FrameType.Open;
        case "data":
            return FrameType.Data;
        case "window":
            return FrameType.Window;
        case "fin":
            return FrameType.Fin;
        case "reset":
            return FrameType.Reset;
    }
}

function decodeUtf8(bytes: Uint8Array, label: string): string {
    try {
        return decoder.decode(bytes);
    } catch {
        throw protocolError(`${label} must be valid UTF-8.`);
    }
}

function assertStreamId(value: number): void {
    if (!Number.isInteger(value) || value <= 0 || value > UINT32_MAX) {
        throw protocolError("Frame streamId must be a positive u32.");
    }
}

function assertPositiveU32(value: number, label: string): void {
    if (!Number.isInteger(value) || value <= 0 || value > UINT32_MAX) {
        throw protocolError(`${label} must be a positive u32.`);
    }
}

function assertU16(value: number, label: string): void {
    if (!Number.isInteger(value) || value <= 0 || value > UINT16_MAX) {
        throw protocolError(`${label} must be a positive u16.`);
    }
}

function assertPacketSize(size: number, maxPacketSize: number): void {
    if (size > maxPacketSize) {
        throw protocolError(
            `Frame payload exceeds ${maxPacketSize} bytes.`,
            "protocol.frameTooLarge",
        );
    }
}

function readU16(bytes: Uint8Array, offset: number): number {
    return view(bytes).getUint16(offset, false);
}

function readU32(bytes: Uint8Array, offset: number): number {
    return view(bytes).getUint32(offset, false);
}

function writeU16(bytes: Uint8Array, offset: number, value: number): void {
    view(bytes).setUint16(offset, value, false);
}

function writeU32(bytes: Uint8Array, offset: number, value: number): void {
    view(bytes).setUint32(offset, value, false);
}

function view(bytes: Uint8Array): DataView {
    return new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
}

function protocolError(message: string, code = "protocol.invalidFrame"): Error {
    return createError({ code: code as ErrorCode, message, retryable: false });
}

File: crates/devshell-worker/src/transport/frame/codec.rs
use std::convert::TryFrom;

const PACKET_HEADER_SIZE: usize = 4;
const FRAME_HEADER_SIZE: usize = 6;
const OPEN_FIXED_SIZE: usize = 6;
const WINDOW_PAYLOAD_SIZE: usize = 4;
const RESET_CODE_SIZE: usize = 2;

pub const FRAME_PROTOCOL_VERSION: u8 = 1;
pub const FRAME_MAX_DATA_SIZE: usize = 64 * 1024;
pub const TRANSPORT_MAX_FRAME_SIZE: usize = 16 * 1024 * 1024;

pub const RESET_UNSUPPORTED_SERVICE: u16 = 1;
pub const RESET_SERVICE_FAILED: u16 = 3;
#[cfg(test)]
pub const RESET_CANCELLED: u16 = 4;

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Frame {
    Open {
        stream_id: u32,
        receive_window: u32,
        service: String,
        metadata: Vec<u8>,
    },
    Data {
        stream_id: u32,
        data: Vec<u8>,
    },
    Window {
        stream_id: u32,
        credit_delta: u32,
    },
    Fin {
        stream_id: u32,
    },
    Reset {
        stream_id: u32,
        code: u16,
        message: String,
    },
}

impl Frame {
    pub fn stream_id(&self) -> u32 {
        match self {
            Self::Open { stream_id, .. }
            | Self::Data { stream_id, .. }
            | Self::Window { stream_id, .. }
            | Self::Fin { stream_id }
            | Self::Reset { stream_id, .. } => *stream_id,
        }
    }
}

#[derive(Debug)]
pub struct FrameDecoder {
    buffer: Vec<u8>,
    max_frame_size: usize,
}

impl Default for FrameDecoder {
    fn default() -> Self {
        Self::new(TRANSPORT_MAX_FRAME_SIZE)
    }
}

impl FrameDecoder {
    pub fn new(max_frame_size: usize) -> Self {
        Self {
            buffer: Vec::new(),
            max_frame_size,
        }
    }

    pub fn is_empty(&self) -> bool {
        self.buffer.is_empty()
    }

    pub fn push(&mut self, bytes: &[u8]) -> Result<Vec<Frame>, String> {
        if !bytes.is_empty() {
            self.buffer.extend_from_slice(bytes);
        }
        let mut frames = Vec::new();
        loop {
            if self.buffer.len() < PACKET_HEADER_SIZE {
                break;
            }
            let body_len = read_u32(&self.buffer, 0)? as usize;
            if body_len > self.max_frame_size {
                return Err(format!(
                    "Frame payload exceeds {} bytes.",
                    self.max_frame_size
                ));
            }
            let packet_len = PACKET_HEADER_SIZE
                .checked_add(body_len)
                .ok_or_else(|| "Frame packet length overflow.".to_string())?;
            if self.buffer.len() < packet_len {
                break;
            }
            let packet = self.buffer[..packet_len].to_vec();
            self.buffer.drain(..packet_len);
            frames.push(decode_frame(&packet)?);
        }
        Ok(frames)
    }
}

pub fn encode_frame(frame: &Frame) -> Result<Vec<u8>, String> {
    let stream_id = frame.stream_id();
    if stream_id == 0 {
        return Err("Frame streamId must be a positive u32.".to_string());
    }

    let (frame_type, payload) = match frame {
        Frame::Open {
            receive_window,
            service,
            metadata,
            ..
        } => {
            if *receive_window == 0 {
                return Err("OPEN receive window must be a positive u32.".to_string());
            }
            let service_bytes = service.as_bytes();
            if service_bytes.is_empty() {
                return Err("OPEN service must be a non-empty UTF-8 name.".to_string());
            }
            let service_len = u16::try_from(service_bytes.len())
                .map_err(|_| "OPEN service exceeds u16 length.".to_string())?;
            let mut payload =
                Vec::with_capacity(OPEN_FIXED_SIZE + service_bytes.len() + metadata.len());
            payload.extend_from_slice(&receive_window.to_be_bytes());
            payload.extend_from_slice(&service_len.to_be_bytes());
            payload.extend_from_slice(service_bytes);
            payload.extend_from_slice(metadata);
            (0x01, payload)
        }
        Frame::Data { data, .. } => {
            if data.len() > FRAME_MAX_DATA_SIZE {
                return Err(format!("DATA payload exceeds {FRAME_MAX_DATA_SIZE} bytes."));
            }
            (0x02, data.clone())
        }
        Frame::Window { credit_delta, .. } => {
            if *credit_delta == 0 {
                return Err("WINDOW credit must be a positive u32.".to_string());
            }
            (0x03, credit_delta.to_be_bytes().to_vec())
        }
        Frame::Fin { .. } => (0x04, Vec::new()),
        Frame::Reset { code, message, .. } => {
            if *code == 0 {
                return Err("RESET code must be a positive u16.".to_string());
            }
            let message = message.as_bytes();
            let mut payload = Vec::with_capacity(RESET_CODE_SIZE + message.len());
            payload.extend_from_slice(&code.to_be_bytes());
            payload.extend_from_slice(message);
            (0x05, payload)
        }
    };

    let body_len = FRAME_HEADER_SIZE
        .checked_add(payload.len())
        .ok_or_else(|| "Frame body length overflow.".to_string())?;
    if body_len > TRANSPORT_MAX_FRAME_SIZE {
        return Err(format!(
            "Frame payload exceeds {TRANSPORT_MAX_FRAME_SIZE} bytes."
        ));
    }
    let body_len_u32 =
        u32::try_from(body_len).map_err(|_| "Frame body exceeds u32 length.".to_string())?;
    let mut output = Vec::with_capacity(PACKET_HEADER_SIZE + body_len);
    output.extend_from_slice(&body_len_u32.to_be_bytes());
    output.push(FRAME_PROTOCOL_VERSION);
    output.push(frame_type);
    output.extend_from_slice(&stream_id.to_be_bytes());
    output.extend_from_slice(&payload);
    Ok(output)
}

pub fn decode_frame(packet: &[u8]) -> Result<Frame, String> {
    if packet.len() < PACKET_HEADER_SIZE {
        return Err("Frame packet header is incomplete.".to_string());
    }
    let body_len = read_u32(packet, 0)? as usize;
    if body_len > TRANSPORT_MAX_FRAME_SIZE {
        return Err(format!(
            "Frame payload exceeds {TRANSPORT_MAX_FRAME_SIZE} bytes."
        ));
    }
    if packet.len() != PACKET_HEADER_SIZE + body_len {
        return Err("Frame packet length does not match payload length.".to_string());
    }
    let body = &packet[PACKET_HEADER_SIZE..];
    if body.len() < FRAME_HEADER_SIZE {
        return Err("Frame header is incomplete.".to_string());
    }
    let version = body[0];
    if version != FRAME_PROTOCOL_VERSION {
        return Err(format!("Unsupported Frame protocol version {version}."));
    }
    let frame_type = body[1];
    let stream_id = read_u32(body, 2)?;
    if stream_id == 0 {
        return Err("Frame streamId must be a positive u32.".to_string());
    }
    let payload = &body[FRAME_HEADER_SIZE..];

    match frame_type {
        0x01 => decode_open(stream_id, payload),
        0x02 => {
            if payload.len() > FRAME_MAX_DATA_SIZE {
                return Err(format!("DATA payload exceeds {FRAME_MAX_DATA_SIZE} bytes."));
            }
            Ok(Frame::Data {
                stream_id,
                data: payload.to_vec(),
            })
        }
        0x03 => {
            if payload.len() != WINDOW_PAYLOAD_SIZE {
                return Err("WINDOW payload must be exactly 4 bytes.".to_string());
            }
            let credit_delta = read_u32(payload, 0)?;
            if credit_delta == 0 {
                return Err("WINDOW credit must be a positive u32.".to_string());
            }
            Ok(Frame::Window {
                stream_id,
                credit_delta,
            })
        }
        0x04 => {
            if !payload.is_empty() {
                return Err("FIN payload must be empty.".to_string());
            }
            Ok(Frame::Fin { stream_id })
        }
        0x05 => {
            if payload.len() < RESET_CODE_SIZE {
                return Err("RESET payload is incomplete.".to_string());
            }
            let code = read_u16(payload, 0)?;
            if code == 0 {
                return Err("RESET code must be a positive u16.".to_string());
            }
            let message = String::from_utf8(payload[RESET_CODE_SIZE..].to_vec())
                .map_err(|_| "RESET message must be valid UTF-8.".to_string())?;
            Ok(Frame::Reset {
                stream_id,
                code,
                message,
            })
        }
        _ => Err(format!("Unknown Frame type {frame_type}.")),
    }
}

fn decode_open(stream_id: u32, payload: &[u8]) -> Result<Frame, String> {
    if payload.len() < OPEN_FIXED_SIZE {
        return Err("OPEN payload is incomplete.".to_string());
    }
    let receive_window = read_u32(payload, 0)?;
    if receive_window == 0 {
        return Err("OPEN receive window must be a positive u32.".to_string());
    }
    let service_len = read_u16(payload, 4)? as usize;
    if service_len == 0 {
        return Err("OPEN service must not be empty.".to_string());
    }
    let service_end = OPEN_FIXED_SIZE
        .checked_add(service_len)
        .ok_or_else(|| "OPEN service length overflow.".to_string())?;
    if service_end > payload.len() {
        return Err("OPEN service length exceeds payload length.".to_string());
    }
    let service = String::from_utf8(payload[OPEN_FIXED_SIZE..service_end].to_vec())
        .map_err(|_| "OPEN service must be valid UTF-8.".to_string())?;
    Ok(Frame::Open {
        stream_id,
        receive_window,
        service,
        metadata: payload[service_end..].to_vec(),
    })
}

fn read_u16(bytes: &[u8], offset: usize) -> Result<u16, String> {
    let end = offset
        .checked_add(2)
        .ok_or_else(|| "u16 offset overflow.".to_string())?;
    let source = bytes
        .get(offset..end)
        .ok_or_else(|| "u16 field is incomplete.".to_string())?;
    Ok(u16::from_be_bytes([source[0], source[1]]))
}

fn read_u32(bytes: &[u8], offset: usize) -> Result<u32, String> {
    let end = offset
        .checked_add(4)
        .ok_or_else(|| "u32 offset overflow.".to_string())?;
    let source = bytes
        .get(offset..end)
        .ok_or_else(|| "u32 field is incomplete.".to_string())?;
    Ok(u32::from_be_bytes([
        source[0], source[1], source[2], source[3],
    ]))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn hex(bytes: &[u8]) -> String {
        bytes.iter().map(|byte| format!("{byte:02x}")).collect()
    }

    #[test]
    fn matches_typescript_frame_v1_wire_vectors() {
        let vectors = [
            (
                Frame::Open {
                    stream_id: 1,
                    receive_window: 256,
                    service: "network.tcp".to_string(),
                    metadata: vec![0xaa, 0xbb],
                },
                "0000001901010000000100000100000b6e6574776f726b2e746370aabb",
            ),
            (
                Frame::Data {
                    stream_id: 2,
                    data: b"abc".to_vec(),
                },
                "00000009010200000002616263",
            ),
            (
                Frame::Window {
                    stream_id: 3,
                    credit_delta: 65_536,
                },
                "0000000a01030000000300010000",
            ),
            (Frame::Fin { stream_id: 4 }, "00000006010400000004"),
            (
                Frame::Reset {
                    stream_id: 5,
                    code: RESET_CANCELLED,
                    message: "stop".to_string(),
                },
                "0000000c010500000005000473746f70",
            ),
        ];

        for (frame, expected) in vectors {
            let encoded = encode_frame(&frame).expect("encode frame");
            assert_eq!(hex(&encoded), expected);
            assert_eq!(decode_frame(&encoded).expect("decode frame"), frame);
        }
    }

    #[test]
    fn decoder_restores_split_and_coalesced_frames() {
        let first = encode_frame(&Frame::Data {
            stream_id: 1,
            data: b"one".to_vec(),
        })
        .expect("encode first");
        let second = encode_frame(&Frame::Fin { stream_id: 1 }).expect("encode second");
        let mut joined = first.clone();
        joined.extend_from_slice(&second);
        let mut decoder = FrameDecoder::default();
        assert!(
            decoder
                .push(&joined[..3])
                .expect("partial header")
                .is_empty()
        );
        assert!(
            decoder
                .push(&joined[3..first.len() - 1])
                .expect("partial body")
                .is_empty()
        );
        assert_eq!(
            decoder.push(&joined[first.len() - 1..]).expect("finish"),
            vec![
                Frame::Data {
                    stream_id: 1,
                    data: b"one".to_vec(),
                },
                Frame::Fin { stream_id: 1 },
            ]
        );
        assert!(decoder.is_empty());
    }
}
